R/register_handlers.R

```diff
@@ -0,0 +1,135 @@
+#'
+#' Use `TEAL.LOG_MUFFLE` environmental variable or `teal.log_muffle` R option to optionally
+#' control recover strategies. If `TRUE` (non-default value) then the handler will jump to muffle
+#' restart for a given type and doesn't continue. This will essentially suppress a given condition
+#' type.
+#' Applicable for message (`"muffleMessage"`) and warning (`"muffleWarning"`) types only.
+#'
+#' @note Registering handlers is forbidden within `tryCatch()` or `withCallingHandlers()`.
+#' Because of this, handlers are registered only if it is possible.
+#'
+#' @seealso [globalCallingHandlers()]
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' register_handlers("teal.logger")
+#' # see the outcome
+#' globalCallingHandlers()
+#' }
+register_handlers <- function(namespace, package = namespace) {
+  if (register_handlers_possible()) {
+    register_handler_type(namespace = namespace, package = package, type = "message")
+    register_handler_type(namespace = namespace, package = package, type = "warning")
+    register_handler_type(namespace = namespace, package = package, type = "error")
+  }
+
+  invisible(NULL)
+}
+
+register_handler_type <- function(
+    namespace,
+    package,
+    type = c("error", "warning", "message")) {
+  match.arg(type)
+  if (!(is.character(namespace) && length(namespace) == 1)) {
+    stop("namespace argument must be a single string.")
+  }
+  if (!(namespace %in% logger::log_namespaces())) {
+    stop("namespace argument must be a pre-registered logger namespace.")
+  }
+  if (!(is.character(package) && length(package) == 1)) {
+    stop("package argument must be a single string.")
+  }
+
+  logger_fun <- switch(type,
+    error = logger::log_error,
+    warning = logger::log_warn,
+    message = logger::log_info
+  )
+
+  # avoid re-registering the same handler
+  gch <- globalCallingHandlers()[names(globalCallingHandlers()) == type]
+  if (
+    length(gch) > 0 &&
+      any(sapply(gch, attr, "type") == "teal.logger_handler" & sapply(gch, attr, "package") == package)
+  ) {
+    return(invisible(NULL))
+  }
+
+  # create a handler object
+  # loop through the call stack and if a function from a given package is found - log the message to the namespace
+  handler_fun <- function(m) {
+    i <- 0L
+    while (i <= sys.nframe()) {
+      if (isNamespaceLoaded(package) && identical(environment(sys.function(i)), getNamespace(package))) {
+        msg <- m$message
+        if (type %in% c("error", "warning") && !is.null(m$call)) {
+          msg <- sprintf("In %s: %s", sQuote(paste0(format(m$call), collapse = "")), msg)
+        }
+        logger_fun(msg, namespace = namespace)
+        if (type == "message" && isTRUE(get_val("TEAL.LOG_MUFFLE", "teal.log_muffle", FALSE))) {
+          invokeRestart("muffleMessage")
+        }
+        if (type == "warning" && isTRUE(get_val("TEAL.LOG_MUFFLE", "teal.log_muffle", FALSE))) {
+          invokeRestart("muffleWarning")
+        }
```

Comment:
hey @gogonzo, `"muffleMessage"` is possible! I was wrong!

However I am not too sure what should be the default value. Currently it's repeating.

(with some dummy testing functions not pushed)
```
r$> setwd("teal.logger")

r$> devtools::load_all()
ℹ Loading teal.logger

r$> message(1)
1

r$> warning(1)
Warning message:
1 

r$> test_message()
[INFO] 2024-03-18 18:25:04.5634 pid:11047 token:[] teal.logger I am a message
I am a message

r$> test_warning()
[WARN] 2024-03-18 18:25:08.3314 pid:11047 token:[] teal.logger In ‘test_warning()’: I am a warning
Warning message:
In test_warning() : I am a warning

r$> withr::with_options(list("teal.log_muffle" = TRUE), test_message())
[INFO] 2024-03-18 18:26:15.5920 pid:11047 token:[] teal.logger I am a message

r$> withr::with_options(list("teal.log_muffle" = TRUE), test_warning())
[WARN] 2024-03-18 18:26:21.1893 pid:11047 token:[] teal.logger In ‘test_warning()’: I am a warning
```

Reply:
Probably because of not muffling - we have duplicated entries. They are duplicated even twice (three entries). Let's muffle them by default


![image](https://github.com/insightsengineering/teal.logger/assets/6959016/2d16f0e3-6c05-4f7e-8c34-23bab5583ee6)

Reply:
ok I can change the default but that would imply that we would have to turn it off during test setup because all of expect-warnings type of things - I'm happy to do that.

The repeating issue requires more thinking though. It's a direct consequence that each package has its own logger namespace. Then the question is: where should "Initializing tm_xyz" log entry go? For sure it has to go to tmg logger namespace because that's the package where it comes from. 
Should it go for teal namespace as well? This is not super clear to me. 
Yes because it would be convenient for devs to have just one logger namespace to analyse and that could be easily the teal one. A follow-up question could be: why do we have other namespaces then? Yes also because it is somewhat coming from teal - but indirectly.
No because that means duplicated entries. If the appender-fun is console (which is a default and should remain so) then experience is even worse.

I think I would be able to implement either - the Q is what we actually want?

What do you think about the above?

Reply:
I feel we reached the conclusion to do muffle restart and not duplicate. Resolving this thread.